Sources/Private/Experimental/Animations/GradientAnimations.swift

```diff
@@ -0,0 +1,129 @@
+// Created by Cal Stephens on 1/7/22.
+// Copyright © 2022 Airbnb Inc. All rights reserved.
+
+import QuartzCore
+
+extension GradientRenderLayer {
+
+  // MARK: Internal
+
+  /// Adds gradient-related animations to this layer, from the given `GradientFill`
+  func addAnimations(for gradientFill: GradientFill, context: LayerAnimationContext) {
+    // We have to set `colors` to a non-nil value with some valid number of colors
+    // for the color animation below to have any effect
+    colors = .init(
+      repeating: CGColor.rgb(0, 0, 0),
+      count: gradientFill.numberOfColors)
+
+    addAnimation(
+      for: .colors,
+      keyframes: gradientFill.colors.keyframes,
+      value: { colorComponents in
+        gradientFill.colorConfiguration(from: colorComponents).map { $0.color }
+      },
+      context: context)
+
+    addAnimation(
+      for: .locations,
+      keyframes: gradientFill.colors.keyframes,
+      value: { colorComponents in
+        gradientFill.colorConfiguration(from: colorComponents).map { $0.location }
+      },
+      context: context)
+
+    switch gradientFill.gradientType {
+    case .linear:
+      addLinearGradientAnimations(for: gradientFill, context: context)
+    case .radial:
+      addRadialGradientAnimations(for: gradientFill, context: context)
+    case .none:
+      break
+    }
+  }
+
+  // MARK: Private
+
+  private func addLinearGradientAnimations(
+    for gradientFill: GradientFill,
+    context: LayerAnimationContext)
+  {
+    type = .axial
+
+    addAnimation(
+      for: .startPoint,
+      keyframes: gradientFill.startPoint.keyframes,
+      value: { absoluteStartPoint in
+        percentBasedPointInBounds(from: absoluteStartPoint.pointValue)
+      },
+      context: context)
+
+    addAnimation(
+      for: .endPoint,
+      keyframes: gradientFill.endPoint.keyframes,
+      value: { absoluteEndPoint in
+        percentBasedPointInBounds(from: absoluteEndPoint.pointValue)
+      },
+      context: context)
+  }
+
+  private func addRadialGradientAnimations(
+    for gradientFill: GradientFill,
+    context _: LayerAnimationContext)
+  {
+    type = .radial
+
+    // To draw the correct gradients, we have to derive a custom `endPoint`
+    // relative to the `startPoint` value. Since calculating the `endPoint`
+    // at any given time requires knowing the current `startPoint`,
+    // we can't allow them to animate separately.
+    guard
+      gradientFill.startPoint.keyframes.count == 1,
+      gradientFill.endPoint.keyframes.count == 1
+    else {
+      fatalError("Animating radial gradient `endPoint` point is currently unsupported")
+    }
+
+    let absoluteStartPoint = gradientFill.startPoint.keyframes.first!.value.pointValue
+    let absoluteEndPoint = gradientFill.endPoint.keyframes.first!.value.pointValue
```

Comment:
nit: Why not do a guard let above? Seems like we check count and then force unwrap

Reply:
Good point, updated this to `[0]`